File: companies/LinkedIn/40.combination-sum-ii.go
package leetcode

import "sort"

// CombinationSum2_40 backtracks over sorted candidates, using each index once.
// At each recursion depth, equal values after the first are skipped to prevent
// duplicate combinations.
//
// Time: O(2^n*n) including output copies. Space: O(n).
func CombinationSum2_40(candidates []int, target int) [][]int {
	sort.Ints(candidates)
	ans := [][]int{}
	path := []int{}
	var dfs func(int, int)
	dfs = func(start, remain int) {
		if remain == 0 {
			ans = append(ans, append([]int(nil), path...))
			return
		}
		prevSet := false
		prev := 0
		for i := start; i < len(candidates); i++ {
			v := candidates[i]
			if prevSet && v == prev {
				continue
			}
			if v > remain {
				break
			}
			path = append(path, v)
			dfs(i+1, remain-v)
			path = path[:len(path)-1]
			prev, prevSet = v, true
		}
	}
	dfs(0, target)
	return ans
}

File: companies/Redfin/40.combination-sum-ii.cpp
#include <algorithm>
#include <cctype>
#include <climits>
#include <cmath>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    vector<vector<int>> combinationSum2(vector<int>& candidates, int target) {
        /*
        Approach:
        Sort candidates so duplicates are adjacent. Backtracking advances to
        i+1 because each number can be used once. At the same recursion depth,
        skip equal values after the first to avoid duplicate combinations.

        Complexity: O(2^n * n) worst-case time including copies, O(n) stack.
        */
        sort(candidates.begin(), candidates.end());
        vector<vector<int>> ans;
        vector<int> path;
        function<void(int, int)> dfs = [&](int start, int remain) {
            if (remain == 0) {
                ans.push_back(path);
                return;
            }
            for (int i = start; i < (int)candidates.size() && candidates[i] <= remain; ++i) {
                if (i > start && candidates[i] == candidates[i - 1]) continue;
                path.push_back(candidates[i]);
                dfs(i + 1, remain - candidates[i]);
                path.pop_back();
            }
        };
        dfs(0, target);
        return ans;
    }
};
